File: projects/14-media-transport/src/media.rs
//! A synthetic constant-bitrate media source — **fully wired**, not a vertical.
//!
//! It emits fixed-size "access units" at a frame rate, a keyframe every GOP, with a 90 kHz
//! RTP timestamp advancing per frame. It is deliberately *not* a real encoder: it produces
//! byte buffers shaped like access units so the sender pipeline (packetize → pace → send)
//! has something to carry before you point a real `ffmpeg -f rtp` feed or camera at it.
//!
//! This is the moral equivalent of project 13's wired `LiveRegistry`: plumbing the vertical
//! work plugs into, given to you so the scaffold runs end-to-end once you fill the todos.

use std::time::Duration;

use bytes::Bytes;
use tokio::time::{interval, Interval, MissedTickBehavior};

use crate::rtp::H264_CLOCK_RATE;

/// One synthetic access unit handed to the packetizer.
pub struct Frame {
    /// Encoded bytes (here: filler sized to hit the target bitrate).
    pub data: Bytes,
    /// Whether this frame is a keyframe (GOP boundary).
    pub keyframe: bool,
    /// 90 kHz RTP timestamp for this frame.
    pub rtp_timestamp: u32,
}

/// A CBR frame generator paced by a tokio interval.
pub struct SyntheticSource {
    ticker: Interval,
    frame_bytes: usize,
    ticks_per_frame: u32,
    gop: u32,
    frame_index: u64,
}

impl SyntheticSource {
    /// `fps` frames/sec, `kbps` target video rate, keyframe every `gop` frames.
    pub fn new(fps: u32, kbps: u32, gop: u32) -> Self {
        let fps = fps.max(1);
        let gop = gop.max(1);
        let mut ticker = interval(Duration::from_micros(1_000_000 / fps as u64));
        // Under load we'd rather skip a tick than try to catch up in a burst.
        ticker.set_missed_tick_behavior(MissedTickBehavior::Skip);
        // bytes/frame = (kbps * 1000 / 8) / fps.
        let frame_bytes = ((kbps.max(1) as usize) * 1000 / 8) / fps as usize;
        Self {
            ticker,
            frame_bytes: frame_bytes.max(1),
            ticks_per_frame: H264_CLOCK_RATE / fps,
            gop,
            frame_index: 0,
        }
    }

    /// Await the next frame's tick and produce it.
    pub async fn next_frame(&mut self) -> Frame {
        self.ticker.tick().await;
        let keyframe = self.frame_index.is_multiple_of(self.gop as u64);
        let rtp_timestamp = (self.frame_index.wrapping_mul(self.ticks_per_frame as u64)) as u32;
        let frame = Frame {
            data: Bytes::from(vec![0u8; self.frame_bytes]),
            keyframe,
            rtp_timestamp,
        };
        self.frame_index = self.frame_index.wrapping_add(1);
        frame
    }
}

File: projects/14-media-transport/src/media_transport/media.py
"""A synthetic constant-bitrate media source — **fully wired**, not a vertical.

It emits fixed-size "access units" at a frame rate, a keyframe every GOP, with a
90 kHz RTP timestamp advancing per frame. It is deliberately *not* an encoder:
it produces byte buffers shaped like access units so the sender pipeline
(packetize → pace → send) has something to carry before you point a real
`ffmpeg -f rtp` feed or a camera at it.

This is the moral equivalent of project 13's wired registry — plumbing the
vertical work plugs into, given to you so the scaffold runs end to end the
moment you fill in the todos.

## Pacing a generator on an event loop

The Rust used `tokio::interval` with `MissedTickBehavior::Skip`. asyncio has no
interval primitive, so the cadence is hand-rolled against a monotonic deadline,
and the "skip" behaviour has to be written out — which is worth doing once,
because the naive version is a bug that compounds.

`await asyncio.sleep(1 / fps)` in a loop does not give you `fps` frames a
second. It gives you one frame every `1/fps` **plus however long the body took**
plus the loop's scheduling latency, so the stream drifts slower and slower and
the RTP timestamps — which advance by a fixed tick count per frame — drift away
from wall-clock time. Over a five-minute boss fight that is a visible desync,
and it looks exactly like a jitter-buffer bug.

Tracking an absolute next-deadline fixes the drift. Skipping deadlines that have
already passed (rather than firing them back to back to "catch up") is the other
half: under load, a burst of frames emitted at once is the last thing a
congestion controller needs, and a frame whose moment has passed is not worth
sending late.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass

from .rtp import H264_CLOCK_RATE, TIMESTAMP_MASK

__all__ = ["Frame", "SyntheticSource"]


@dataclass(frozen=True, slots=True)
class Frame:
    """One synthetic access unit handed to the packetizer."""

    data: bytes
    """Encoded bytes — here, filler sized to hit the target bitrate."""
    keyframe: bool
    """True at a GOP boundary. V3 and V4 both care: a keyframe is the frame
    whose loss is worth recovering hardest, and the burst that most stresses the
    pacer."""
    rtp_timestamp: int
    """90 kHz media timestamp for this frame, shared by all its packets."""


class SyntheticSource:
    """A CBR frame generator paced against a monotonic deadline.

    `bitrate_bps` is re-read from the congestion controller between frames by
    the sender loop, so the synthetic stream actually responds to V4 instead of
    ignoring it — which is what makes the convergence criterion observable
    without a real encoder in the loop.
    """

    __slots__ = ("_frame_index", "_fps", "_gop", "_next_deadline", "_ticks_per_frame")

    def __init__(self, fps: int, gop: int) -> None:
        self._fps = max(1, fps)
        self._gop = max(1, gop)
        self._ticks_per_frame = H264_CLOCK_RATE // self._fps
        self._frame_index = 0
        self._next_deadline = 0.0

    @property
    def frame_interval(self) -> float:
        """Seconds between frames — the interval the pacer spreads a frame over."""
        return 1.0 / self._fps

    async def next_frame(self, bitrate_bps: int) -> Frame:
        """Wait for the next frame's moment, then produce it at `bitrate_bps`.

        The bitrate is a parameter rather than constructor state because it
        changes: the sender passes the congestion controller's current target on
        every call, so a backed-off estimate immediately produces smaller
        frames. That is the feedback loop closing.
        """
        now = time.monotonic()
        if self._next_deadline == 0.0:
            self._next_deadline = now
        # Skip, don't catch up: deadlines already in the past are abandoned
        # rather than fired back to back. See the module docstring.
        while self._next_deadline <= now:
            self._next_deadline += self.frame_interval
        await asyncio.sleep(self._next_deadline - now)

        frame_bytes = max(1, (bitrate_bps // 8) // self._fps)
        frame = Frame(
            data=bytes(frame_bytes),
            keyframe=self._frame_index % self._gop == 0,
            rtp_timestamp=(self._frame_index * self._ticks_per_frame) & TIMESTAMP_MASK,
        )
        self._frame_index += 1
        return frame
